File: libs/nox-ecs/examples/six_dof.rs
use impeller2::component::Component;
use impeller2_wkt::{Color, Line3d, Material, Mesh, Panel, Viewport};
use nox::{
    tensor, ArrayRepr, SpatialForce, SpatialInertia, SpatialMotion, SpatialTransform, Vector3,
};
use nox_ecs::{six_dof::*, Integrator, Query, World, WorldExt, WorldPos};

fn gravity(pos: Query<(WorldPos, Inertia, Force)>) -> Query<Force> {
    const G: f64 = 6.649e-11;
    let big_m: f64 = 1.0 / G;
    pos.map(|world_pos: WorldPos, inertia: Inertia, force: Force| {
        let mass = inertia.0.mass();
        let r = world_pos.0.linear();
        let mask: Vector3<f64> = tensor![1.0, 1.0, 0.0].into();
        let r = r * mask;
        let norm = r.clone().norm();
        let force = force.0
            + SpatialForce::from_linear(
                -r / (norm.clone() * norm.clone() * norm) * G * big_m * mass,
            );
        Force(force)
    })
    .unwrap()
}

fn main() {
    stellarator::run(|| async {
        tracing_subscriber::fmt::init();
        let mut world = World::default();
        let shape = world.insert_shape(Mesh::sphere(0.1), Material::color(1.0, 0.0, 0.0));

        let shape_b = world.insert_shape(Mesh::sphere(0.1), Material::color(0.5, 1.0, 1.0));

        let a = world
            .spawn(Body {
                pos: WorldPos(SpatialTransform {
                    inner: tensor![1.0, 0.0, 0.0, 0.0, -1.0, 0.0, 0.0].into(),
                }),
                vel: WorldVel(SpatialMotion {
                    inner: tensor![0.0, 0.0, 0.0, 0.0, -1.0, 0.0].into(),
                }),
                accel: WorldAccel(SpatialMotion {
                    inner: tensor![0.0, 0.0, 0.0, 0.0, 0.0, 0.0].into(),
                }),
                force: Force(SpatialForce {
                    inner: tensor![0.0, 0.0, 0.0, 0.0, 0.0, 0.0].into(),
                }),
                mass: Inertia(SpatialInertia {
                    inner: tensor![1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0].into(),
                }),
            })
            .insert(shape)
            .id();

        let b = world
            .spawn(Body {
                pos: WorldPos(SpatialTransform {
                    inner: tensor![1.0, 0.0, 0.0, 0.0, -1.0, 0.0, 0.0].into(),
                }),
                vel: WorldVel(SpatialMotion {
                    inner: tensor![0.0, 0.0, 0.0, 0.0, 1.0, 2.0].into(),
                }),
                accel: WorldAccel(SpatialMotion {
                    inner: tensor![0.0, 0.0, 0.0, 0.0, 0.0, 0.0].into(),
                }),
                force: Force(SpatialForce {
                    inner: tensor![0.0, 0.0, 0.0, 0.0, 0.0, 0.0].into(),
                }),
                mass: Inertia(SpatialInertia {
                    inner: tensor![1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0].into(),
                }),
            })
            .insert(shape_b)
            .id();
        let panel = world.insert_asset(Panel::Viewport(Viewport {
            ..Default::default()
        }));
        world.spawn(panel);
        let asset = world.insert_asset(Line3d {
            entity: b,
            component_id: <WorldPos<ArrayRepr>>::COMPONENT_ID,
            index: [4, 5, 6],
            line_width: 10.0,
            color: Color::YOLK,
            perspective: true,
        });
        world.spawn(asset);

        let asset = world.insert_asset(Line3d {
            entity: a,
            component_id: <WorldPos<ArrayRepr>>::COMPONENT_ID,
            index: [4, 5, 6],
            line_width: 10.0,
            color: Color::MINT,
            perspective: true,
        });
        world.spawn(asset);

        let exec = world
            .builder()
            .tick_pipeline(six_dof(|| gravity, Integrator::Rk4))
            .build()
            .unwrap();
        let client = nox::Client::cpu().unwrap();
        let exec = exec.compile(client).unwrap();
        nox_ecs::impeller2_server::Server::new(
            elodin_db::Server::new("./test", "0.0.0.0:2240".parse().unwrap()).unwrap(),
            exec,
        )
        .run()
        .await
        .unwrap()
    })
}

File: libs/nox-py/examples/six_dof.py
from jax import numpy as np
from jax.numpy import linalg as la
from elodin import *

TIME_STEP = 1.0 / 120.0


def gravity_impl(pos, inertia, force):
    G = 6.67430e-11
    big_m = 1 / G
    mass = inertia.mass()
    r = pos.linear()
    norm = la.norm(r)
    f = -r / norm * norm * norm * G * big_m * mass
    return Force.from_linear(f)


@system
def gravity(q: Query[WorldPos, Inertia, Force]) -> Query[Force]:
    return q.map(Force, gravity_impl)


w = WorldBuilder()
w.spawn(
    Body(
        world_pos=WorldPos.from_linear(np.array([0.0, 0.0, 0.01])),
        world_vel=WorldVel.from_linear(np.array([0.0, 0.0, 0.0])),
        inertia=Inertia.from_mass(1.0),
        pbr=w.insert_asset(Pbr(Mesh.sphere(1.0), Material.color(25.3, 18.4, 1.0))),
    )
)
w.spawn(
    Body(
        world_pos=WorldPos.from_linear(np.array([5.0, 0.0, 0.0])),
        world_vel=WorldVel.from_linear(np.array([0.0, 0.0, 10.0])),
        inertia=Inertia.from_mass(1.0),
        pbr=w.insert_asset(Pbr(Mesh.sphere(0.2), Material.color(1.0, 1.0, 1.0))),
    )
)
w.spawn(
    Body(
        world_pos=WorldPos.from_linear(np.array([8.0, 0.0, 0.0])),
        world_vel=WorldVel.from_linear(np.array([0.0, 0.0, 24.0])),
        inertia=Inertia.from_mass(2.0),
        pbr=w.insert_asset(Pbr(Mesh.sphere(0.3), Material.color(1.0, 1.0, 1.0))),
    )
)
sys = six_dof(TIME_STEP, gravity)
sim = w.run(sys, TIME_STEP)
